File: 3kyu/esolang_interpreters_3.js
function interpreter(code, iterations, width, height) {
  console.log(code, iterations, width, height);
  const grid = new Array(height);
  for (let i = 0; i < height; i++) {
    grid[i] = [];
    for (let j = 0; j < width; j++) {
      grid[i][j] = 0;
    }
  }
  let i = 0;
  let j = 0;
  let cp = 0;
  let ip = 0;
  while (ip < iterations && cp < code.length) {
    const c = code[cp];
    switch (c) {
      case 'n':
        i = (i - 1 + height) % height;
        break;
      case 'e':
        j = (j + 1) % width;
        break;
      case 's':
        i = (i + 1) % height;
        break;
      case 'w':
        j = (j - 1 + width) % width;
        break;
      case '*':
        grid[i][j] = grid[i][j] ? 0 : 1;
        break;
      case '[':
        if (!grid[i][j]) {
          let ct = 0;
          while (!(code[cp] == ']' && ct == 1)) {
            if (code[cp] == ']') ct--;
            else if (code[cp] == '[') ct++;
            cp++;
          }
        }
        break;
      case ']':
        if (grid[i][j]) {
          let ct = 0;
          while (!(code[cp] == '[' && ct == 1)) {
            if (code[cp] == ']') ct++;
            else if (code[cp] == '[') ct--;
            cp--;
          }
        }
        break;
      default:
        ip--;
        break;
    }
    cp++;
    ip++;
  }

  return grid.map(row => row.join('')).join('\r\n');
}

File: 3kyu/esolang_interpreters_3.go
package kata

import (
	"fmt"
	"strings"
)

func Interpreter(code string, iterations, width, height int) string {
	fmt.Println(code)
	grid := make([][]int, height)
	for i := range grid {
		grid[i] = make([]int, width)
	}

	i := 0
	j := 0
	cp := 0
	ip := 0

	for ip < iterations && cp < len(code) {
		switch code[cp] {
		case 'n':
			i = (i - 1 + height) % height
		case 'e':
			j = (j + 1) % width
		case 's':
			i = (i + 1) % height
		case 'w':
			j = (j - 1 + width) % width
		case '*':
			if grid[i][j] == 1 {
				grid[i][j] = 0
			} else {
				grid[i][j] = 1
			}
		case '[':
			if grid[i][j] == 0 {
				ct := 0
				for !(code[cp] == ']' && ct == 1) {
					if code[cp] == ']' {
						ct--
					} else if code[cp] == '[' {
						ct++
					}
					cp++
				}
			}
		case ']':
			if grid[i][j] == 1 {
				ct := 0
				for !(code[cp] == '[' && ct == 1) {
					if code[cp] == ']' {
						ct++
					} else if code[cp] == '[' {
						ct--
					}
					cp--
				}
			}
		default:
			ip--
		}
		cp++
		ip++
	}

	ans := make([]string, height)
	for i, r := range grid {
		for _, c := range r {
			ans[i] += fmt.Sprintf("%d", c)
		}
	}
	return strings.Join(ans, "\r\n")
}
